### DISCUSSION TITLE
How to work functions with multiple promises?

### DISCUSSION THREAD

**Discussion by mogarick** [statelyai/xstate#2265] [Q&A]
Given a scenario where I have a step where I need to execute business logic that requires calling more than one micro service, how can I manage that with XState invoke given the fact that it will resolve as soon as the first promise called resolves?

example function to call using invoke:

``` js
async function myAsyncFn(params) {
  try {
  // some business logic code

  let foo =  await firstAsyncFn(params);

  //some business logic with foo value

  await secondAsyncFn(otherparams);

  //some additional business logic
 } catch(e) {
   //some compensation logic
   
  throw new Error("A wrapped error message");
 }

}
```

example state fragment using invoke 

``` js
{
  states: {
    myState: {
              entry: "someSetup",
              //NOTE: I also noticed invoke is called before entry. That's weird for me. I assumed entry always should get executed first.
              invoke: {
                id: "someService",
                src: "myAsyncFn",
                onDone: {
                  target: "anotherState",
                },
                onError: {
                  target: "failedSate",
                },
              },
              exit: "someExitCleanup",
            },
  }
}
```

What happens here is that as soon as `firstAsyncFn` resolves the entry action gets executed even though `secondAsyncFn` and the other logic have not finished execution. It also triggers `onTransition` listeners after this first async call resolves. I was expecting any other action, such as `entry` action,  to be executed only after invoke call to `myAsyncFn`  has finished completely either succeeding or failing (in fact I was expecting `entry` to be executed before invoke but that's not how it works as I could notice).

So it appears that scenarios involving more than one async call inside a function are not supported. Am I right? Is there a suggested work around on this subject?

Thank you in advance.

**Comment by davidkpiano** [comment]
✏️ Please create a [CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed) reproduction of this. Thanks!

**Reply by mogarick** [reply]
Ok thank you. I'll do it and post the link here ASAP.
Just to clarify, The invoked function is not aborted after the first call. It continues executing until finished. But what happens is that after the first promise resolves, the `entry` function is executed and the same for `onTransition`, even though the second promise and the rest of the async fn is still executing and finishes later and either goes to `onDone` or `onError`.

**Reply by mogarick** [reply]
Hi again,

I think I mistakenly said the execution of `entry` and `onTransition` continues after the first promise resolves.  It doesn't. What happens is that as soon as the first promise is called, `entry` and `onTransition` execute while first promise is still pending.

This is expected according to the use of `Promise.resolve(src)` I saw in the interpreter code but I'm not sure if that is the intention.
I would expect the call to `entry` to be done before invoke and the `onTransition` to be called either after or before `entry` depending on when a transition to a state can be considered to be finished.

Here is the codesandbox as requested.

Thank you in advance for the attention.

https://codesandbox.io/s/xstate-example-template-vanilla-forked-6rk9d?file=/src/index.js:378-381

**Reply by davidkpiano** [reply]
> I would expect the call to `entry` to be done before invoke and the `onTransition` to be called either after or before `entry` depending on when a transition to a state can be considered to be finished.

You should not rely on the order of `entry` actions and `invoke`; `entry` is fire-and-forget, so its effects must be considered _in isolation_. If you want the invocation to be dependent on the action, then the action should be part of the invocation, _not_ a separate action.

Can you detail your use-case a bit more? What are you doing in `entry` that the invocation depends on?

However, according to [the algorithm](https://www.w3.org/TR/scxml/#mainEventLoop) it seems like invocations should happen _after_ `entry` actions. cc. @Andarist to verify

**Reply by mogarick** [reply]
> Can you detail your use-case a bit more? What are you doing in entry that the invocation depends on?

I'm experimenting with Xstate to migrate an app I built using some custom state-machine-like logic.
I'm thinking on using `entry` for actions like rendering a "loading" UI component and invoke to load data. When loading finishes `onDone` will target a "loaded" state in which its respective `entry` action would be the rendering/presentation of the UI component.

> However, according to the algorithm it seems like invocations should happen after entry actions. cc. @Andarist to verify

It would be excellent if the execution is done in order as `entry` hints it's something that occurs as soon as state is entered and `exit` as soon as state is being abandoned.

Now with respect to `onTransition` I think it should be triggered after the machine has finished making all the internal actions related to that state, such as `actions` called upon the event that triggered the state change, `entry` or `invoke`. Once the machine has "settled" on such state then it should "notify" to its listeners that it's now in a new state. 
Does it work that way or what's the current way it works?

**Reply by davidkpiano** [reply]
@mogarick To simplify things, you can use tags to provide rendering hints: https://github.com/davidkpiano/xstate/releases/tag/xstate%404.19.0

Again, order implies coupling, and actions are supposed to be isolated effects. If you want an effect to happen in the beginning of an invocation, I suggest moving it inside the invocation instead of as a separate action.

Some example code would be helpful, if you want to provide a CodeSandbox?

**Reply by github0013** [reply]
@mogarick 
Can you not simply move and add the `stateA`'s entry action to `starting`'s exit?

entering -> (actions) -> starting -> exiting -> (actions) -> entering -> (actions/invoke) -> stateA -> ...

Doesn't it make sure the specific action is done when you are at the state of `stateA`?

**Reply by mogarick** [reply]
Sorry for the delay @davidkpiano. I'll check later about tags. Thank you.
Back to the main subject of this question, is there a reason for the behavior that occurs when invoking a Promise that internally includes more than one await?

As I described it before, when invoking the promise, as soon as the first await is hit, the runtime switches to the execution of external code and after it finishes that then it continues with the execution of the rest of the invoke method call and doesn't switch away until the last internal await finishes and then resolves the invocation.

So the current behavior is this:

```
invoke->promise function->first await call->external code execution->first await finished->more promise function code->2nd await call->2nd await finished->more promise function code->resolved->[onDone|onErrror]
```

Is this the expected behavior per the current implementation of the invoke promise feature?

This behavior is kind of weird to me because the "external code execution" time varies with the complexity of the stateMachine going from a 2-4 ms to 2000+ ms in my local env with an mbpro 2018 with 32GB RAM (last macOS update). But if I execute the same code in codesandbox it only varies from 2ms to 6 ms.

I created a sandbox with 2 machines, one simpler that the other. On line 323 you can play with both of them to see what I mean. Look for `externalCodeExecution:#####.####### ms` on the console.
If you could execute the same code on local env you'd should see the big differences in execution time with respect to the condesandbox env.

https://codesandbox.io/s/silly-morning-mi9j1?file=/src/index.ts

I hope you, @Andarist or someone from the XState team can look at it and tell me if it is some kind of bug or if I'm doing something really wrong here.

Thank you again for the help.

**Reply by davidkpiano** [reply]
> Is this the expected behavior per the current implementation of the invoke promise feature?

If I understand correctly, you're asking if the expected behavior is that the _entire_ `async` function is awaited to be resolved before calling `onDone`/`onError` transitions, right? That is the expected behavior, regardless of XState - you wait for the entire thing to be resolved.

What behavior are you expecting exactly?

**Reply by mogarick** [reply]
> If I understand correctly, you're asking if the expected behavior is that the entire async function is awaited to be resolved before calling onDone/onError transitions, right? That is the expected behavior, regardless of XState - you wait for the entire thing to be resolved.
What behavior are you expecting exactly?

I was expecting the call to `entry` and to the `onTransition` fn to be executed sequentially before the call to the invoked fn or after it resolves.
But what it does is it begins executing the invoked fn, executes `entry` actions, executes `onTransition` listener fn and finally gets back to the invoked fn, finishes executing its code and then resolves/fails.

And the other thing I don't get is what other code it executes when it encounters the first `await` and switch to the execution of `entry` actions and `onTransition` listener because for `machineDefinition` and `machineDefinition2` the action code to execute, external to the invoked fn, is supposed to be the same, except for more states declared in `machineDefinition2` that are not  supposed to be executed during the call to the invoked fn. But even so the execution time varies widely as described in my previous message.

I hope I better described the situation. 

You can check the codesandbox I linked to in such previous message.

Thank you again for the attention.
